### DISCUSSION TITLE
Sharing common config across projects, with package instances for dev/prod

### DISCUSSION THREAD

**Discussion by chriswarbo** [cue-lang/cue#1415] [Q&A]
Hello, I currently have a lot of AWS CloudFormation yaml which I'd like to simplify and validate using cue. I have multiple projects, each with their own git repo and `template.yaml` file (which are largely copy/pasted at the moment). I would like to transition to each repo having a `template.cue` file instead, which imports common definitions from a central repo. I've got this working quite nicely, by essentially running `cd /path/to/common/module && cue export --out yaml - < /path/to/template.cue` (in fact I'm using Nix to check out the repos and run that command).

For example:

```
$ cd /path/to/common/module && cue export --out yaml < /path/to/project/template.cue
AWSTemplateFormatVersion: "2010-09-09"
Description: Example stack
Resources:
  MyStream:
    Type: AWS::Kinesis::Stream
    Properties:
      RetentionPeriodInHours: 24

$ cat /path/to/project/template.cue
import "example.com/cloudformation"
cloudformation.#Stack
Description: "Example stack"

Resources: {
  MyStream: {
    cloudformation.#KinesisStream,
    Properties: {
      RetentionPeriodInHours: 24
    }
  }
}

$ cat /path/to/common/module/cue.mod/module.cue 
module: "example.com"

$ cat /path/to/common/module/cloudformation/stacks.cue
package cloudformation
import (
	"struct"
)

#Stack: {
	AWSTemplateFormatVersion: "2010-09-09"
	Description:              string

	// Needs at least one #Resource
	Resources: struct.MinFields(1) & {...}
	Resources: [Name=string]: #Resource
}

#Resource: {
	Type: string
	Properties: struct.MinFields(1)
}

$ cat /path/to/common/module/cloudformation/kinesis.cue 
package cloudformation

#KinesisStream: {
	#Resource,
	Type: "AWS::Kinesis::Stream",
	Properties: {
		Name?: string,
		ShardCount?: int
		RetentionPeriodHours?: int
	}
}
```

This works well, but now I'd like to extend it to specify different values/constraints for different deployment environments (dev, prod, etc.). I've tried to do this with cue "instances", like this:

```
$ cat /path/to/common/module/dev/cloudformation/constraints.cue
package cloudformation

// Keep things small and cheap on dev
#KinesisStream: Properties: ShardCount: 1

$ cd /path/to/common/modules/dev && cue export --out yaml < /path/to/project/template.cue
AWSTemplateFormatVersion: "2010-09-09"
Description: Example stack
Resources:
  MyStream:
    Type: AWS::Kinesis::Stream
    Properties:
      RetentionPeriodInHours: 24
```

The contents of subdirectories (like dev) seems to be ignored, e.g. there's no `ShardCount: 1` in the yaml.

I've tried a bunch of different options, e.g. `cue export . - < ....`, moving the `.cue` files out of their `cloudformation` directories, etc. It seems I can either export the dev instance's definitions on their own, or I can export the project's config merged with the top-level common definitions, but I can't export the project's config merged with the top-level and dev instance definitions.

Any help would be appreciated. Note that I'd like to keep each project in their separate repos; I don't want the config repo to contain project-specific details.

**Comment by brk3** [comment]
Hi, there are a couple of issues I see here.

The first is that cuelang `Instances` only combine files that are both in the same package, and ancestors of the current directory. In your sample, directories `cloudformation` and `dev` are siblings, so this won't work.

Next, I'm not sure it makes sense to be running `cue export` from the `cloudformation` package. We should aim to build an `Instance` in our consuming package, and customize from there. Try this...

* Move `cue.mod` up from `common` to the root level. This makes both `common` and `project` part of the same `module`. This is just to save the hassle of manually vendoring `common` under the `project` `cue.mod/pkg` dir, as Cue doesn't yet have dependency management. In your example, only `common` has a `cue.mod`, meaning imports won't work in `project`. I'm thinking this might be what led you to run `cue export` from `common`?

Next, move the `dev` constraint under `project`, and change it's package statement to `project`. This way when you run `cue export` from here, it will be picked up as part of the same `Instance`.

Finally, in `project/template.cue`, don't hardcode the reference to `#KinesisStream` from the `cloudformation` package. Instead, create a placeholder variable, which can default to `cloudformation.#KinesisStream`, but gets superseded with `project.#KinesisStream` when `cue export` is ran from within `project/dev/cloudformation`:

```
package project

import "example.com/common/cloudformation"

#KinesisStream: _ | *cloudformation.#KinesisStream

cloudformation.#Stack
Description: "Example stack"

Resources: {
        MyStream: {
                #KinesisStream
                Properties: {
                        RetentionPeriodInHours: 24
                }
        }
}
```

Notes:
* This stuff is kind of hard to explain in text, so I've uploaded a sample repo here: https://github.com/brk3/cuelang-instances-demo (directory 'ex1' reflects your question, 'ex2' reflects my answer).

* If you wanted, you could probably also keep the `constraints.cue` in the common package under a sub package, and then have a `dev` config in `project` that controls which type ultimately gets imported. Might be a nice follow up exercise :)
 
* I'm quite new to cuelang packages so this may not be 100% correct, but it's something to go on. Hopefully some more experienced cuelang folk can verify what I've written!

**Reply by chriswarbo** [reply]
Thanks for taking the time to help :) I'll give this a try; although there are a couple of details which will make life a bit trickier:

 1. I'm using multiple independent git repos. The examples of `/path/to/common/module` and `/path/to/package` are meant to represent arbitrary directories, where those git repos happen to be checked-out. In particular, there is no "parent folder" in which to put `cue.mod`. I can solve this using Nix, by constructing such folders as-needed.
 2. I would prefer to pipe `template.cue` into the `cue` tool, rather than have it read off disk; since that lets me generate the .cue programatically. I don't think this is a deal-breaker, since the things I'm generating could probably be achieved using cue's tag feature instead :)
